contractcourt/htlc_incoming_contest_resolver.go

```diff
@@ -441,19 +490,78 @@ func (h *htlcIncomingContestResolver) SupplementState(_ *channeldb.OpenChannel)
 }
 
 // decodePayload (re)decodes the hop payload of a received htlc.
-func (h *htlcIncomingContestResolver) decodePayload() (*hop.Payload, error) {
+func (h *htlcIncomingContestResolver) decodePayload() (*hop.Payload,
+	[]byte, error) {
 
 	onionReader := bytes.NewReader(h.htlc.OnionBlob)
 	iterator, err := h.OnionProcessor.ReconstructHopIterator(
 		onionReader, h.htlc.RHash[:],
 	)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
+	}
+
+	payload, err := iterator.HopPayload()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Transform onion blob for the next hop.
+	var onionBlob [lnwire.OnionPacketSize]byte
+	buf := bytes.NewBuffer(onionBlob[0:0])
+	err = iterator.EncodeNextHop(buf)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	return iterator.HopPayload()
+	return payload, onionBlob[:], nil
 }
 
 // A compile time assertion to ensure htlcIncomingContestResolver meets the
 // ContractResolver interface.
 var _ htlcContractResolver = (*htlcIncomingContestResolver)(nil)
+
+type interceptedForward struct {
+	packet      *InterceptedPacket
+	resolveChan chan lntypes.Preimage
+}
+
+func newInterceptedForward(
+	packet *InterceptedPacket,
+	resolveChan chan lntypes.Preimage) *interceptedForward {
+
+	return &interceptedForward{
+		resolveChan: resolveChan,
+		packet:      packet,
+	}
+}
+
+// Packet returns the intercepted htlc packet.
+func (f *interceptedForward) Packet() InterceptedPacket {
+	return *f.packet
+}
+
+// Resume resumes the default behavior as if the packet was not intercepted.
+func (f *interceptedForward) Resume() error {
+	// In the on-chain stage, there is nothing to resume.
+
+	return nil
+}
+
+// Fail forward a failed packet to the switch.
+func (f *interceptedForward) Fail() error {
+	// TODO: Think about whether we want a quick fail, or keep waiting for a
```

Comment:
I think we should honor the fail if the interceptor wants it? Otherwise people may end up with accounting issues?

Reply:
Returning an error now and documented the race condition that would occur.

If the interceptor fails after the to-be-implemented `fail_height`, I think it is reasonable to return an error.

Reply:
Although, yes, channels can also go to chain for another reason, with the htlc still far from expiry. So I guess the interceptor needs to be prepared for receiving an error at any time.

Maybe the fundamental problem here is that in the off-chain resolution flow, the preimage flow is isolated per htlc. Then in on-chain, we use a global preimage cache. This prevents a reliable fail action.